// storybook-zeplin/dist/components/OverlayPanel.js
import React, { useCallback, useReducer } from "react";
import { styled } from "@storybook/theming";
import OverlayPortal from "./OverlayPortal";
import OverlayButtons from "./OverlayButtons";
import OverlayImage from "./OverlayImage";
const initialState = {
    showOverlay: false,
    lockOverlay: false,
    showDifference: false,
    overlayScaling: 0.5,
    opacity: 1,
};
const OverlayPanel = ({ imageUrl }) => {
    const [state, setState] = useReducer((state, newState) => ({
        ...state,
        ...newState,
    }), initialState);
    const { showOverlay, lockOverlay, showDifference, overlayScaling, opacity, } = state;
    const selectScaling = useCallback((event) => {
        setState({ overlayScaling: event.target.value });
    }, []);
    const toggleOverlay = () => {
        setState({ showOverlay: !showOverlay });
    };
    const toggleLock = () => {
        setState({ lockOverlay: !lockOverlay });
    };
    const toggleDifference = () => {
        setState({ showDifference: !showDifference });
    };
    const updateOpacity = (event) => {
        setState({ opacity: event.currentTarget.value });
    };
    return (React.createElement(React.Fragment, null,
        React.createElement(OverlayButtons, { overlayIsOpen: showOverlay, overlayIsLocked: lockOverlay, showDifference: showDifference, onToggleOverlay: toggleOverlay, onToggleLock: toggleLock, onToggleDifference: toggleDifference }),
        showOverlay && (React.createElement(OverlayOptions, null,
            "Opacity",
            React.createElement(Input, { type: "range", min: "0", max: "1", step: "0.01", value: opacity, onChange: updateOpacity }),
            "Scaling",
            React.createElement(Select, { onChange: selectScaling, value: overlayScaling },
                React.createElement("option", { value: 0.5 }, "x0.5"),
                React.createElement("option", { value: 1 }, "x1"),
                React.createElement("option", { value: 2 }, "x2")))),
        showOverlay && (React.createElement(OverlayPortal, null,
            React.createElement(OverlayImage, { url: imageUrl, opacity: opacity, isLocked: lockOverlay, scaling: overlayScaling, showDifference: showDifference })))));
};
export default OverlayPanel;
const Select = styled.select `
    margin-left: 15px;
`;
const Input = styled.input `
  margin-right: 30px;
  margin-left: 15px;
  width: 100%;
`;
const OverlayOptions = styled.div `
    margin-left: 15px;
    display: flex;
    align-items: center;
`;
