// @zeplin/cli/dist/src/tasks/ui/select-component.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.failed = exports.completed = exports.select = exports.noMatchingComponent = exports.skippedSelection = exports.foundNoComponents = exports.retrieving = exports.initial = void 0;
const chalk_1 = __importDefault(require("chalk"));
exports.initial = () => ({
    text: "Select a Zeplin component"
});
exports.retrieving = () => ({
    text: "Retrieving Zeplin components..."
});
exports.foundNoComponents = (ctx) => ({
    text: `No components found in ${ctx.selectedResource.name}`
});
exports.skippedSelection = (ctx) => {
    var _a;
    return ({
        text: `Selected ${chalk_1.default.cyan((_a = ctx.selectedComponents) === null || _a === void 0 ? void 0 : _a.map(c => c.name)[0])}`,
        subtext: `You provided --component-id=${ctx.cliOptions.componentId}`
    });
};
exports.noMatchingComponent = (ctx) => ({
    text: `Could not find matching component ${chalk_1.default.cyan(ctx.cliOptions.componentId)}`
});
exports.select = (ctx) => ({
    text: "Select a Zeplin component...",
    subtext: `${ctx.selectedResource.type}: ${ctx.selectedResource.name}`
});
exports.completed = (ctx) => {
    var _a;
    return ({
        text: `Selected Zeplin ${ctx.selectedComponents.length > 0 ? "components" : "component"}:`,
        subtext: `${(_a = ctx.selectedComponents) === null || _a === void 0 ? void 0 : _a.map(c => c.name).join("\n")}`
    });
};
exports.failed = () => ({
    text: "Failed to select a Zeplin component"
});
