// @zeplin/cli/dist/src/tasks/select-file.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.selectFile = void 0;
const inquirer_1 = __importDefault(require("inquirer"));
const inquirer_fuzzy_path_1 = __importDefault(require("inquirer-fuzzy-path"));
const fs = __importStar(require("fs-extra"));
const path_1 = __importDefault(require("path"));
const task_1 = require("../util/task");
const ui = __importStar(require("./ui/select-file"));
const error_1 = require("../util/task/error");
const messages_1 = require("../messages");
inquirer_1.default.registerPrompt("fuzzypath", inquirer_fuzzy_path_1.default);
const excludedPaths = [
    "node_modules",
    ".git"
    // TODO: add more excluded paths for other types of projects
];
function createFileMetadata(filepath) {
    const absolutePath = path_1.default.resolve(filepath);
    return {
        name: path_1.default.basename(filepath),
        extension: path_1.default.extname(filepath),
        path: path_1.default.relative(process.cwd(), filepath),
        absolutePath
    };
}
const checkResourceFlags = (ctx, task) => __awaiter(void 0, void 0, void 0, function* () {
    if (ctx.cliOptions.filename) {
        const absolutePath = path_1.default.resolve(ctx.cliOptions.filename);
        if (!(yield fs.pathExists(absolutePath))) {
            throw new error_1.TaskError(ui.fileNotFound);
        }
        ctx.file = createFileMetadata(absolutePath);
        task.complete(ctx, ui.completed);
    }
});
const select = (ctx) => __awaiter(void 0, void 0, void 0, function* () {
    const { selection } = yield inquirer_1.default.prompt([{
            type: "fuzzypath",
            name: "selection",
            itemType: "file",
            rootPath: ".",
            message: messages_1.selectComponentFilePrompt,
            excludePath: (p) => {
                for (const excludedPath of excludedPaths) {
                    if (p.startsWith(excludedPath)) {
                        return true;
                    }
                }
                return false;
            }
        }]);
    ctx.file = createFileMetadata(selection);
});
exports.selectFile = new task_1.Task({
    steps: [
        checkResourceFlags,
        task_1.pauseSpinningAndExecuteTask(select),
        task_1.transitionTo(ui.completed)
    ],
    initial: ui.initial
});
