// @zeplin/cli/dist/src/service/project-type/project-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSupportedProjectType = exports.supportedProjectTypes = exports.ProjectType = void 0;
const js_dependency_1 = require("./matchers/js-dependency");
const storybook_1 = require("./storybook");
const storybook_2 = require("./configurators/storybook");
var ProjectType;
(function (ProjectType) {
    ProjectType["REACT"] = "react";
    ProjectType["PREACT"] = "preact";
    ProjectType["ANGULAR"] = "angular";
    ProjectType["VUE"] = "vue";
    ProjectType["HAS_STORYBOOK"] = "storybook";
    ProjectType["UNKNOWN"] = "unknown";
    // TODO: Add other detectable frameworks
})(ProjectType = exports.ProjectType || (exports.ProjectType = {}));
exports.supportedProjectTypes = [
    {
        type: ProjectType.REACT,
        matcher: js_dependency_1.jsDependencyMatcherFactory(["react", "react-native"]),
        installPackages: [
            "@zeplin/cli-connect-react-plugin"
        ]
    },
    {
        type: ProjectType.PREACT,
        matcher: js_dependency_1.jsDependencyMatcherFactory(["preact"]),
        installPackages: [
            "@zeplin/cli-connect-react-plugin"
        ]
    },
    {
        type: ProjectType.ANGULAR,
        matcher: js_dependency_1.jsDependencyMatcherFactory(["@angular/core"]),
        installPackages: [
            "@zeplin/cli-connect-angular-plugin"
        ]
    },
    {
        type: ProjectType.VUE,
        matcher: js_dependency_1.jsDependencyMatcherFactory(["vue", "nuxt"]),
        installPackages: [
            "zeplin-cli-connect-plugin-vue"
        ]
    },
    {
        type: ProjectType.HAS_STORYBOOK,
        matcher: js_dependency_1.jsDependencyMatcherFactory(storybook_1.supportedStorybookFrameworks),
        installPackages: [
            "@zeplin/cli-connect-storybook-plugin"
        ],
        configurator: storybook_2.storybookConfiguratorFactory()
    }
];
function getSupportedProjectType(projectType) {
    return exports.supportedProjectTypes.find(spt => spt.type === projectType);
}
exports.getSupportedProjectType = getSupportedProjectType;
