// @zeplin/cli/dist/src/commands/connect/server/connect-dev-server.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectDevServer = void 0;
const chalk_1 = __importDefault(require("chalk"));
const express_1 = __importDefault(require("express"));
const http_status_codes_1 = require("http-status-codes");
const errors_1 = require("../../../errors");
const logger_1 = __importDefault(require("../../../util/logger"));
class ConnectDevServer {
    constructor(connectedBarrels) {
        this.connectedBarrels = [];
        this.stopped = false;
        this.connections = [];
        this.connectedBarrels = connectedBarrels;
    }
    getConnectedComponents(barrelId) {
        const found = this.connectedBarrels.find(connectedBarrel => connectedBarrel.projects.find(pid => pid === barrelId) ||
            connectedBarrel.styleguides.find(stid => stid === barrelId));
        return found ? found.connectedComponents : null;
    }
    updateConnectedBarrels(connectedBarrels) {
        this.connectedBarrels = connectedBarrels;
    }
    start(port) {
        var _a;
        if ((_a = this.server) === null || _a === void 0 ? void 0 : _a.listening) {
            return Promise.resolve(this.server);
        }
        const app = express_1.default();
        // CORS
        app.use((_req, res, next) => {
            res.header("Access-Control-Allow-Origin", "*");
            res.header("Access-Control-Allow-Headers", "X-Requested-With, Origin");
            res.header("Access-Control-Allow-Methods", "GET");
            next();
        });
        app.get("/:type/:barrelId/connectedcomponents", (req, res) => {
            const { barrelId } = req.params;
            const connectedComponents = this.getConnectedComponents(barrelId);
            return res.status(http_status_codes_1.OK).json({ connectedComponents });
        });
        return new Promise((resolve, reject) => {
            this.server = app.listen(port)
                .on("listening", () => {
                logger_1.default.debug(`Started dev server on port ${port}`);
                logger_1.default.info(chalk_1.default.green(`Development server is started.`));
                resolve(this.server);
            })
                .on("error", (err) => {
                if (err.code === "EADDRINUSE") {
                    reject(new errors_1.CLIError(`Port ${port} is already in use.`));
                }
            })
                .on("connection", connection => {
                this.connections.push(connection);
                connection.on("close", () => (this.connections = this.connections.filter(curr => curr !== connection)));
            });
        });
    }
    stop() {
        logger_1.default.debug("Stopping dev server.");
        this.stopped = true;
        this.connections.forEach(conn => conn.end());
        return new Promise((resolve) => {
            var _a;
            (_a = this.server) === null || _a === void 0 ? void 0 : _a.close(() => {
                logger_1.default.debug("Stopped dev server.");
                resolve();
            });
        });
    }
    listen(port) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.start(port);
            return new Promise((resolve) => {
                process.on("SIGINT", () => __awaiter(this, void 0, void 0, function* () {
                    yield this.stop();
                    resolve();
                }));
            });
        });
    }
}
exports.ConnectDevServer = ConnectDevServer;
